File: BoundaryEstimation.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local BoundaryEstimation = torch.class('pcl.BoundaryEstimation', pcl)

local func_by_type = {}

local function init()

  local BoundaryEstimation_method_names = {
    'new',
    'delete',
    'setInputCloud',
    'setInputNormals',
    'setIndices',
    'setAngleThreshold',
    'getAngleThreshold',
    'setSearchMethod_Octree',
    'setSearchMethod_KdTree',
    'setKSearch',
    'getKSearch',
    'setRadiusSearch',
    'getRadiusSearch',
    'compute',
    'computeIndices'
  }
  for k,v in pairs(utils.type_key_map) do
    func_by_type[k] = utils.create_typed_methods("pcl_BoundaryEstimation_TYPE_KEY_", BoundaryEstimation_method_names, v)
  end    
end

init()

function BoundaryEstimation:__init(pointType)
  self.pointType = pcl.pointType(pointType or pcl.PointXYZ)
  self.f = func_by_type[self.pointType]
  self.o = self.f.new()
end

function BoundaryEstimation:cdata()
  return self.o
end

function BoundaryEstimation:setInputCloud(cloud)
  self.f.setInputCloud(self.o, cloud:cdata())
end

function BoundaryEstimation:setInputNormals(normals)
  self.f.setInputNormals(self.o, normals:cdata())
end

function BoundaryEstimation:setIndices(indices)
  self.f.setIndices(self.o, indices:cdata())
end

function BoundaryEstimation:setAngleThreshold(angle)
  self.f.setAngleThreshold(self.o, angle)
end

function BoundaryEstimation:getAngleThreshold()
  return self.f.getAngleThreshold(self.o)
end

function BoundaryEstimation:setSearchMethod(search)
  if torch.isTypeOf(search, pcl.KdTree) then
    self.f.setSearchMethod_KdTree(self.o, search:cdata())
  elseif torch.isTypeOf(search, pcl.Octree) then
    self.f.setSearchMethod_Octree(self.o, search:cdata())
  else
    error("unsupported search method")
  end
end
    
function BoundaryEstimation:setKSearch(k)
  self.f.setKSearch(self.o, k)
end

function BoundaryEstimation:getKSearch()
  return self.f.getKSearch()
end

function BoundaryEstimation:setRadiusSearch(radius)
  self.f.setRadiusSearch(self.o, radius)
end

function BoundaryEstimation:getRadiusSearch()
  return self.f.getRadiusSearch(self.o)
end

function BoundaryEstimation:compute(output)
  if not output then
    output = pcl.PointCloud(pcl.Normal)
  end
  self.f.compute(self.o, output:cdata())
  return output
end

function BoundaryEstimation:computeIndices(output)
  output = output or pcl.Indices()
  self.f.computeIndices(self.o, output:cdata())
  return output
end

File: src/generic/boundary_estimation.cpp
#include "searchwrapper.h"
#include <pcl/features/boundary.h>

#define BoundaryEstimation_ptr pcl::BoundaryEstimation<_PointT, pcl::Normal, pcl::Boundary>::Ptr
#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr
#define BoundaryCloud_ptr pcl::PointCloud<pcl::Boundary>::Ptr

#define KdTree_ptr pcl::KdTreeFLANN<_PointT>::Ptr
#define Octree_ptr pcl::octree::OctreePointCloudSearch<_PointT>::Ptr

PCLIMP(BoundaryEstimation_ptr *, BoundaryEstimation, new)()
{
  return new BoundaryEstimation_ptr(new pcl::BoundaryEstimation<_PointT, pcl::Normal, pcl::Boundary>());
}

PCLIMP(void, BoundaryEstimation, delete)(BoundaryEstimation_ptr *self)
{
  delete self;
}

PCLIMP(void, BoundaryEstimation, setInputCloud)(BoundaryEstimation_ptr *self, PointCloud_ptr *cloud)
{
  (*self)->setInputCloud(*cloud);
}

PCLIMP(void, BoundaryEstimation, setInputNormals)(BoundaryEstimation_ptr *self, Normals_ptr *normals)
{
  (*self)->setInputNormals(*normals);
}

PCLIMP(void, BoundaryEstimation, setIndices)(BoundaryEstimation_ptr *self, Indices_ptr *indices)
{
  (*self)->setIndices(*indices);
}

PCLIMP(void, BoundaryEstimation, setAngleThreshold)(BoundaryEstimation_ptr *self, float angle)
{
  (*self)->setAngleThreshold(angle);
}

PCLIMP(float, BoundaryEstimation, getAngleThreshold)(BoundaryEstimation_ptr *self)
{
  return (*self)->getAngleThreshold();
}

PCLIMP(void, BoundaryEstimation, setSearchMethod_Octree)(BoundaryEstimation_ptr *self, Octree_ptr *octree)
{
  boost::shared_ptr<OctreeSearchWrapper<_PointT> > wrapper(new OctreeSearchWrapper<_PointT>(*octree));
  (*self)->setSearchMethod(wrapper);
}

PCLIMP(void, BoundaryEstimation, setSearchMethod_KdTree)(BoundaryEstimation_ptr *self, KdTree_ptr *kdtree)
{
  boost::shared_ptr<KdTreeSearchWrapper<_PointT> > wrapper(new KdTreeSearchWrapper<_PointT>(*kdtree));
  (*self)->setSearchMethod(wrapper);
}

PCLIMP(void, BoundaryEstimation, setKSearch)(BoundaryEstimation_ptr *self, int k)
{
  (*self)->setKSearch(k);
}

PCLIMP(int, BoundaryEstimation, getKSearch)(BoundaryEstimation_ptr *self)
{
  return (*self)->getKSearch();
}

PCLIMP(void, BoundaryEstimation, setRadiusSearch)(BoundaryEstimation_ptr *self, double radius)
{
  (*self)->setRadiusSearch(radius);
}

PCLIMP(double, BoundaryEstimation, getRadiusSearch)(BoundaryEstimation_ptr *self)
{
  return (*self)->getRadiusSearch();
}

PCLIMP(void, BoundaryEstimation, compute)(BoundaryEstimation_ptr *self, BoundaryCloud_ptr *output)
{
  (*self)->compute(**output);
}

PCLIMP(void, BoundaryEstimation, computeIndices)(BoundaryEstimation_ptr *self, Indices_ptr *indices)
{
  pcl::PointCloud<pcl::Boundary> boundary;
  (*self)->compute(boundary);
  
  // filter indices of points on boundary
  std::vector<int>& result = **indices;
  result.clear();
  for (size_t i = 0; i < boundary.points.size(); ++i)
  {
    if (!boundary[i].boundary_point)
      continue;
    
    result.push_back(i);
  }
}

#undef BoundaryEstimation_ptr
#undef PointCloud_ptr
#undef BoundaryCloud_ptr
#undef KdTree_ptr
#undef Octree_ptr
